### DISCUSSION TITLE
Optional Schema fails Validation

### DISCUSSION THREAD

**Discussion by Shumuu** [react-hook-form/react-hook-form#8616] [Q&A] [answered]
Hello!

I have an Issue with Zod and the hookform-resolvers.

I have a Schema where both fields are optionally, although when one key property is set to "true", the other property is required.

the hookform seemingly uses the wrong schema (from the union) to validate and shows for both fields an Error Message, even though only one field failed validation

To see my problem ...

1. Go to the [Codesandbox](https://codesandbox.io/s/elegant-kalam-isy2wh?file=/src/schema.js)
2. Check `True`
3. Click `Submit`
4. See Error for the fields `test` and `testAmount`.

I guess this issue wouldn't occur when I could use a `discriminatedUnion` but I'd hope for there to be another way.

**Comment by Moshyfawn** [comment, marked as answer]
This is not the RHF / resolvers issue, but rather the way Zod works. From [the Zod's "Discriminated unions" section](https://github.com/colinhacks/zod#discriminated-unions=) of their docs:
> With the basic union method the input is tested against each of the provided "options", and in the case of invalidity, issues for all the "options" are shown in the zod error.

So, the answer is to use [the `discriminatedUnion` operator](https://github.com/colinhacks/zod#discriminated-unions=) instead
